core/primitives/src/shard_layout.rs

```diff
@@ -83,17 +96,195 @@ pub struct ShardLayoutV1 {
+    }
+}
+
+impl fmt::Debug for AccountRange {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let range = format!("{:?}..{:?}", self.start, self.end);
+        f.debug_struct("AccountRange").field("range", &range).finish()
+    }
+}
+
+impl AccountRange {
+    pub fn new_single_shard() -> Self {
+        Self { start: AccountBoundary::Start, end: AccountBoundary::End }
+    }
+
+    pub fn new_start(end: &str) -> Self {
+        Self { start: AccountBoundary::Start, end: AccountBoundary::Middle(end.parse().unwrap()) }
+    }
+
+    pub fn new_end(start: &str) -> Self {
+        Self { start: AccountBoundary::Middle(start.parse().unwrap()), end: AccountBoundary::End }
+    }
+
+    pub fn new_mid(start: &str, end: &str) -> Self {
+        Self {
+            start: AccountBoundary::Middle(start.parse().unwrap()),
+            end: AccountBoundary::Middle(end.parse().unwrap()),
+        }
+    }
+
+    fn contains(&self, account_id: &AccountId) -> bool {
+        match (&self.start, &self.end) {
+            (AccountBoundary::Start, AccountBoundary::End) => true,
+            (AccountBoundary::Start, AccountBoundary::Middle(end)) => account_id < end,
+            (AccountBoundary::Middle(start), AccountBoundary::End) => account_id >= start,
+            (AccountBoundary::Middle(start), AccountBoundary::Middle(end)) => {
+                account_id >= start && account_id < end
+            }
+            (AccountBoundary::End, _) => panic!("invalid account range"),
+            (_, AccountBoundary::Start) => panic!("invalid account range"),
+        }
+    }
+}
+
+/// A mapping from the shard id to the account range of this shard.
+type ShardsAccountRange = BTreeMap<ShardId, AccountRange>;
+
+fn validate_shards_account_range(
+    shards_account_range: &ShardsAccountRange,
+) -> Result<(), ShardLayoutError> {
+    let err =
+        |reason: &str| ShardLayoutError::InvalidShardsAccountRange { reason: reason.to_string() };
+
+    let values = shards_account_range.values().sorted().collect_vec();
+    println!("{:?}", values);
+
+    let (Some(first), Some(last)) = (values.first(), values.last()) else {
+        return Err(err("account range empty"));
+    };
+    if first.start != AccountBoundary::Start {
+        return Err(err("first account range should start with Start"));
+    }
+    if last.end != AccountBoundary::End {
+        return Err(err("last account range should end with End"));
+    }
+
+    for i in 1..values.len() {
+        let prev = values[i - 1];
+        let curr = values[i];
+        if prev.end != curr.start {
+            return Err(err("account ranges should be contiguous"));
+        }
+    }
+
+    Ok(())
+}
+
+/// Making the shard ids non-contiguous.
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ShardLayoutV2 {
+    /// A mapping from the shard id to the account range of this shard.
+    shards_account_range: ShardsAccountRange,
```

Comment:
Why not to leave `boundary_accounts: Vec<AccountId>` and just add new field `shard_ids: Vec<ShardId>`? Then we would have to validate only that `boundary_accounts.len() + 1 == shard_ids.len()` and that `boundary_accounts` is in ascending order. After that, we can construct immediately valid `ShardsAccountRange` for local usage.

Reply:
Yeah, this can be a fine and more simplistic strategy, but I don't feel too strongly about it vs the current implementation.

Reply:
I think either would work but I feel like shard id => range mapping is easier for humans to parse. I'm also considering changing it to range => shard id as that would make the `account_id_to_shard_id` query faster.

Reply:
I agree with human readability argument. But I think readability should be handled by some tool over the shard layout format, like we already do for the latest snapshot of runtime config. I prefer the base primitive to be as concise as possible. For me it's like getting rpc response in json or text - both make sense, but json must be the base format.

If we go with ranges, `start` and `end` are valid accounts so we need some placeholders which are definitely invalid (`$start`?)

Reply:
All right, I'm fine with that, will do tomorrow. 

> If we go with ranges, start and end are valid accounts so we need some placeholders which are definitely invalid ($start?) 

I wouldn't want to use magic values but perhaps I can use Unbounded for those.